lib/os/fdtable.c

```diff
@@ -211,17 +268,46 @@ off_t lseek(int fd, off_t offset, int whence)
 }
 FUNC_ALIAS(lseek, _lseek, off_t);
 
-int ioctl(int fd, unsigned long request, ...)
+int z_impl_sys_ioctl(int fd, unsigned long request, va_list args)
 {
-	va_list args;
-	int res;
-
 	if (_check_fd(fd) < 0) {
 		return -1;
 	}
 
+	return fdtable[fd].vtable->ioctl(fdtable[fd].obj, request, args);
+}
+
+#ifdef CONFIG_USERSPACE
+ssize_t z_vrfy_sys_ioctl(int fd, unsigned long request, va_list args)
```

Comment:
You cannot send a va_list over a system call.
That implies that the kernel would then walk arbitrary stack data that the user provided. This also presumes that a va_list is even something that fits in a register.
Consider also that there is also no way here of validating the arguments provided.

In Linux the underlying system call for ioctl() has three arguments. The userspace code in BIONIC C library looks like:

```
extern "C" int __ioctl(int, int, void *);

int ioctl(int fd, int request, ...) {
  va_list ap;
  va_start(ap, request);
  void* arg = va_arg(ap, void*);
  va_end(ap);
  return __ioctl(fd, request, arg);
}
```

What gets sent to the kernel is a file descriptor, a request ID, and then an untyped pointer to memory.

In order to properly validate, you will need to first validate the file descriptor and then validate the request ID. Once you have validated both of those, we will then know what the exact semantics of the third argument are and can then make the appropriate calls to validate it and copy any data required.

Right now the code is only validating the descriptor. You have to validate everything.

Reply:
> That implies that the kernel would then walk arbitrary stack data that the user provided.

Well, I wouldn't call it "walk the stack", just "access a few words on stack". The biggest risk I see is that va_list doesn't really have to refer to stack, but can be crafted to point to arbitrary memory.

But let's assume it's legitimate reference to stack. Please help me to understand what bad things happen if we don't Z_SYSCALL_MEMORY_READ() this reference in a syscall verify handler. Let's consider that a user stack has just 2 words left and will underflow afterwards, but kernel syscall will want to access 3 words.

I tried to find answer in the docs, and don't see detailed enough description. Actually, some details confusing. I traced Z_SYSCALL_MEMORY_READ to https://docs.zephyrproject.org/latest/guides/porting/arch.html?highlight=arch_buffer_validate#_CPPv420arch_buffer_validatePv6size_ti , 

> Given a memory region, return whether the **current memory management hardware configuration** would allow a user thread to read/write that region.

(Emphasis mine.) Reading such a description, I picture following 2-stage syscall handling sequence:

1. Syscall transfer to kernel mode, but MPU mapping still remains userspace. That's how verify handler is called.
2. A trivial implementation of arch_buffer_validate() thus would just try to read/write each byte in a provided buffer. This would be inefficient, but correct implementation.
3. Only after that, MPU mapping is changed. And possibly changed drastically, like flat mapping is installed, allowing kernel to access entire CPU address space. A more natural solution would be to just enable an MPU region for kernel memory. But then userspace map would still be installed, and then explicit Z_SYSCALL_MEMORY_READ step isn't strictly required.

Is this something resembling reality? It's hard to believe in that, given that verify handler is what calls implementation handler, and there's nothing in there which looks like switching memory map. But then, why arch_buffer_validate() doc refers to "current memory management **hardware** configuration". Hardware configuration by that time can be quite different. And arch_buffer_validate() potentially could be implemented just as series of bounds checks against memory domains configured for calling threads, which has nothing to do with "current hardware configuration" (which could a flat all-addressspace map, again).

 Summing up IMHO there're following issues with https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html :

1. It doesn't specify at which point userspace memory map is switched to kernel-space memory map, which is IMHO important point.
2. It doesn't specify possible examples of what kernel-space memory map have. I understand this is inherently platform-specific, but the talk is exactly about giving a few possible rights, to set *bounds* of expectations right. E.g. can it be full address space enabled for read-write? If yes, say so. Can it be just a userspace map, with a single extra memory region for kernel r/w data enabled? If yes, say so. Note that "full address space" case effectively means that Zephyr userspace support can work with just a single MPU region (for thread stack, + background region to make the rest R/O), and then kernel space map is just disabling MPU - I never read about such frugal requirements).
3. The references to "current hardware configuration" in arch_buffer_validate() are confusing and may not correspond to reality (and thus should be removed).

Thanks for clarifications.